{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c31eddc8-6543-4b15-ba9a-90e215009cd9",
   "metadata": {},
   "source": [
    "# An automated pipeline to create an atlas of in-situ hybridization gene expression data in the adult marmoset brain (ISBI 2023, Poon, C., et al.)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "992f59a1-ab52-4741-b613-cd4b0071a3af",
   "metadata": {},
   "source": [
    "## 3D registration code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47fc2834-8a2b-4542-8b6c-c32af1a3ef99",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "sys.path.append(\"/disk/soft/SLURM/slurm/\")\n",
    "from slurm import slurm_tools\n",
    "import numpy as np\n",
    "import nibabel as nib\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "from PIL import Image\n",
    "import cv2\n",
    "from skimage.transform import resize\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "from IPython.display import clear_output\n",
    "import matplotlib as mpl\n",
    "dpi_default = mpl.rcParams['figure.dpi']\n",
    "\n",
    "def json_read(filename):\n",
    "    if (os.path.isfile(filename)):\n",
    "        with open(filename) as data_file:\n",
    "            data = json.load(data_file)\n",
    "        isfile=True;\n",
    "    else:\n",
    "        data={}\n",
    "        isfile=False;\n",
    "    return data, isfile"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ed7645f-ea90-46ee-92f1-f33175ddb39e",
   "metadata": {},
   "source": [
    "### 1. Align gene images by each RGB channel to the last 2D transform, save as niftis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce5d54dc-a5f2-408a-904b-24af21cc29b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1. aligns the RGB gene niis wrt the last 2D transform, then outputs each aligned RGB image\n",
    "\n",
    "def align_save_gene_rgbs(all_cmp, mid):\n",
    "    print('------in 1 align_save_gene_rgbs')\n",
    "    jobids = []\n",
    "    for gene in all_cmp:\n",
    "\n",
    "        if  \"gene\" in gene:\n",
    "\n",
    "            print('gene:',gene)\n",
    "\n",
    "            # original niftis\n",
    "            niif_ = ch_workdir+\"/\"+mid+\"/\"+gene+\"/img_*.nii.gz\"\n",
    "            files = glob.glob(niif_)\n",
    "            files.sort()\n",
    "            #print('niif_ 1: {}, num files: {}'.format(niif_, len(files)))\n",
    "\n",
    "\n",
    "            # aligned genes\n",
    "            niif_ = ch_workdir+\"/\"+mid+\"/out/\"+gene+\"/2/img/img_*.nii.gz\"\n",
    "            files2 = glob.glob(niif_)\n",
    "            files2.sort()\n",
    "            #print('niif_ 2: {}, num files2: {}'.format(niif_, len(files2)))\n",
    "\n",
    "            # inverse transforms\n",
    "            h5f_ = ch_workdir+\"/\"+mid+\"/out/\"+gene+\"/2/trafo/*InverseComposite.h5\"\n",
    "            files3 = glob.glob(h5f_)\n",
    "            files3.sort()\n",
    "            jobids = []\n",
    "            #print('h5f_: {}, num files3: {}'.format(h5f_, len(files3)))\n",
    "\n",
    "            os.makedirs(workdir+mid+\"/out/\"+gene+\"/2/img/\",exist_ok=True)\n",
    "\n",
    "\n",
    "            for f1,f2,f3,indx in zip(files,files2,files3,range(len(files3))):\n",
    "\n",
    "                for r in [\"R\",\"G\",\"B\"]:\n",
    "                    print(r)\n",
    "                    move = f1.replace(\"img_\",\"img\"+r+\"_\")  # original\n",
    "                    ref = f2                               # aligned\n",
    "                    out = f2.replace(\"img_\",\"img\"+r+\"_\")#.replace(\"charissa\",\"skibbe\")    # does not yet exist\n",
    "                    #print('move: {}, ref: {}, out: {}'.format(move, ref, out))\n",
    "\n",
    "                    cores = \"2\"\n",
    "                    logfolder = \"/\".join(out.split(\"/\")[:-1])+\"/log/\"  # char changed this\n",
    "                    os.makedirs(logfolder,exist_ok=True)\n",
    "\n",
    "                    POSTTRAFO = f3.replace(\"InverseComposite.h5\",\"\")\n",
    "                    #print('logfolder: {}, POSTTRAFO: {}'.format(logfolder, POSTTRAFO))\n",
    "\n",
    "                    SLURM_commands = [\"MOVE=\"+move+\" REF=\"+ref+\" CORES=\"+cores+\" OUT1=\"+out+\"  POSTTRAFO=\"+POSTTRAFO+\" bash apply_2D.sh;\"]\n",
    "                    #SLURM_commands = [\"MOVE=\"+move+\" REF=\"+ref+\" CORES=\"+cores+\" OUT1=\"+OUT1+\"  POSTTRAFO=\"+POSTTRAFO+\"  bash apply_avg3D_02.sh;\"]\n",
    "\n",
    "\n",
    "                    jobid, success = slurm_tools.slurm_submit(SLURM_commands,\n",
    "                        name = \"reg3D\",\n",
    "                        output = logfolder+'/log_2D'+str(indx)+'.out',\n",
    "                        mem = '8GB',\n",
    "                        cores = cores,\n",
    "                        partition =\"ish-adult\", #\"bigmem\",  \n",
    "                        )      \n",
    "\n",
    "                    if not success:\n",
    "                            print(\"could not submit jobs\")\n",
    "                            print(format(jobid))\n",
    "                            print(format(jobids))\n",
    "                            slurm_tools.killall(jobids)\n",
    "                    jobids.append(int(jobid))\n",
    "                slurm_tools.wait_for_jobs(jobids)\n",
    "                clear_output(wait=True)\n",
    "                print('in 1 align_save_gene_rgbs:'+gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa86771c-1534-4766-bdf7-3deb0ad0aa25",
   "metadata": {},
   "source": [
    "### 2. Make 3D sagittal stacks of each gene in each RGB channel, with padding in Z-plane"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75be2f6b-eeb9-445a-910a-06a56a49b543",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 2. make 3D sagittal stacks of each gene (R,B,G), with z-margin (literal z, not z wrt sagittal)\n",
    "# ie the directionality of the slices will be the same as the template backlit image\n",
    "# prep for registration\n",
    "\n",
    "def make_sagittal_stacks(all_cmp,mid):\n",
    "    print('------in 2 make_sagittal_stacks')\n",
    "\n",
    "    for cmp in all_cmp:\n",
    "\n",
    "        if \"backlit\" in cmp or \"gene\" in cmp:\n",
    "        #if  \"gene\" in cmp:  # this was uncommented!!!\n",
    "\n",
    "            # set state\n",
    "            state = 10 if \"backlit\" in cmp or \"seg\" in cmp else 2\n",
    "\n",
    "            print('cmp: {}, state: {}'.format(cmp,state))\n",
    "\n",
    "            # idir is actually the same dir as the 2D registration\n",
    "            if \"seg\" in cmp:\n",
    "                idir = workdir+mid+\"/out_seg/\"+str(cmp)+\"/\"+str(1)+'/img/'\n",
    "            elif \"backlit\" in cmp:\n",
    "                idir = workdir+mid+\"/out/\"+str(cmp)+\"/\"+str(10)+'/img/'\n",
    "            else:\n",
    "                idir = workdir+mid+\"/out/\"+str(cmp)+\"/\"+str(state)+'/img/'\n",
    "            \n",
    "            # if 'backlit' in compartment name, then use the single grayscale image, else use R,G,B\n",
    "            ilist = [\"img_\"] if state == 10 else  [\"img_\",\"imgR_\",\"imgG_\",\"imgB_\"]\n",
    "\n",
    "            for il in ilist:\n",
    "                i_files = glob.glob(idir+\"/\"+il+\"*.nii.gz\")\n",
    "                \n",
    "                i_files.sort()\n",
    "                print('num files: ',len(i_files)) \n",
    "                print('il: ',il)\n",
    "                print('idir: ',idir)\n",
    "\n",
    "                z_margin = 20\n",
    "\n",
    "                for fn,fi in zip(i_files,range(len(i_files))):\n",
    "                    tmp = nib.load(fn).get_fdata(dtype=np.float32)  \n",
    "                    if fi == 0:\n",
    "                        n_files = len(i_files)\n",
    "\n",
    "                        shape = [tmp.shape[1],n_files+2*z_margin,tmp.shape[0]]  \n",
    "\n",
    "                        img3D = np.zeros(shape,dtype=np.uint8)\n",
    "                        print('cmp: {}, n_files: {}, shape: {}, img3D: {}, tmp.max: {}'.format(cmp, n_files, shape, img3D.shape, tmp.max()))\n",
    "\n",
    "\n",
    "                    np.clip(tmp,0,1,tmp) \n",
    "\n",
    "                    # reverse order, rotate images to be in the same direction as template backlit\n",
    "                    img3D[:,n_files-fi-1+z_margin,:] = np.rot90(tmp*255,k=3)\n",
    "\n",
    "\n",
    "                if state != 10:  #\"backlit\" not in cmp:\n",
    "                    assert(img3D[:,:z_margin,:].max()==0)           \n",
    "                    assert(img3D[:,z_margin+n_files:,:].max()==0)\n",
    "                    img3D[:,:z_margin,:] = img3D[:,z_margin,None,:]  \n",
    "                    img3D[:,z_margin+n_files:,:] = img3D[:,z_margin+n_files-1,None,:]  \n",
    "                \n",
    "                if \"seg\" in cmp:\n",
    "                    odir3 = workdir+mid+\"/out_seg/\"+str(cmp)+\"/3D/\"\n",
    "                else:\n",
    "                    odir3 = workdir+mid+\"/out/\"+str(cmp)+\"/3D/\"\n",
    "                os.makedirs(odir3,exist_ok=True)\n",
    "                \n",
    "                print('odir3: {}'.format(odir3))\n",
    "\n",
    "                mat = np.identity(4)\n",
    "                scale = 2.24  \n",
    "                mat[1,1] = 0.6 / scale \n",
    "                mat[0,0] = 0.1 / scale \n",
    "                mat[2,2] = 0.1 / scale  \n",
    "\n",
    "                new_image = nib.Nifti1Image(img3D, affine=mat)\n",
    "                new_image.header[\"cal_min\"] = 0\n",
    "                new_image.header[\"cal_max\"] = 1\n",
    "                new_image.header[\"scl_slope\"] = 1/255.0\n",
    "                new_image.header[\"scl_inter\"] = 0\n",
    "                nib.save(new_image,odir3+il.replace(\"_\",\"\")+\".nii.gz\") \n",
    "\n",
    "    clear_output(wait=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86ac81b3-a552-4240-8620-69883dddf656",
   "metadata": {},
   "source": [
    "### 3. Align backlit images (specific to each marmoset) to the BMCA backlit template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa92f7cb-145f-4344-9fef-45d719b4018e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 3. align backlit with reference backlit brain, save as horizontal stack -- img_2_STPT.nii.gz\n",
    "\n",
    "def align_bl_2_bmcabl(mid):\n",
    "    print('------in 3 align_bl_2_bmcabl:')\n",
    "    jobids = []\n",
    "    cores = \"80\"\n",
    "    odir3 = workdir+mid+\"/out/01_backlit/3D/\" \n",
    "    move = odir3+\"/img.nii.gz\"\n",
    "    ref = workdir + \"../ref/new_ref_bl.nii\"\n",
    "    OUT1 = odir3+\"/img_2_STPT.nii.gz\"\n",
    "    POSTTRAFO = odir3+\"/trafo_01_\"  #\n",
    "    logfolder = odir3+\"/log_3D_bl2bmcabl/\"\n",
    "    os.makedirs(logfolder,exist_ok=True)\n",
    "    os.makedirs(odir3,exist_ok=True)\n",
    "\n",
    "    print('odir3: {}, move: {}, ref: {}, OUT1: {}, POSTTRAFO: {}, logfolder: {}'.format(odir3, move, ref, OUT1, POSTTRAFO, logfolder))\n",
    "\n",
    "    SLURM_commands = [\"MOVE=\"+move+\" REF=\"+ref+\" CORES=\"+cores+\" OUT1=\"+OUT1+\"  POSTTRAFO=\"+POSTTRAFO+\"  bash regme_avg3D_01.sh;\"]\n",
    "\n",
    "\n",
    "    jobid, success = slurm_tools.slurm_submit(SLURM_commands,\n",
    "        name = \"reg3D_bl2bmcabl\",\n",
    "        output = logfolder+'/log_01.out',\n",
    "        mem = '32GB',\n",
    "        cores = cores,\n",
    "        partition = \"ish-adult\",  #\"bigmem\"\n",
    "        )      \n",
    "\n",
    "    print(\"job number: \"+jobid)\n",
    "\n",
    "    if not success:\n",
    "            print(\"could not submit jobs\")\n",
    "            print(format(jobid))\n",
    "            print(format(jobids))\n",
    "            slurm_tools.killall(jobids)\n",
    "    jobids.append(int(jobid))\n",
    "    # break\n",
    "\n",
    "    #print(\"state \"+str(state))\n",
    "    slurm_tools.wait_for_jobs(jobids)  # char commented out\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a4e56a3-e7b7-4aad-951c-5bca94e7a3c7",
   "metadata": {},
   "source": [
    "### 4. Make a backlit template that is larger in the XY plane and smaller in the Z plane"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "657e28e7-a887-41e9-b55a-876e6912c199",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 4. make backlit template that is larger in xy plane and smaller in z plane: new_ref_25mu.nii \n",
    "# only need to make this once\n",
    "\n",
    "if True:\n",
    "    ref_org = nib.load(ref)  # \"ref/new_ref_bl.nii\"\n",
    "    ref_org_d = ref_org.get_fdata(dtype=np.float32)\n",
    "    ref_org_s = ref_org.shape  # (256, 356, 230)\n",
    "    dummy = np.zeros([ref_org_s[0]*4,ref_org_s[1]//8,ref_org_s[2]*4])  # 1024,44,920, (x,z,y)\n",
    "    for a in range(ref_org_s[1]//8):  # 44\n",
    "        #print(a)\n",
    "        dummy[:,a,:] = resize(ref_org_d[:,a*8,:],[ref_org_s[0]*4,ref_org_s[2]*4])  # larger in xy planes and smaller in z plane\n",
    "\n",
    "    dummy = 255*dummy / dummy.max()\n",
    "    affine = ref_org.affine*1\n",
    "    print(affine)\n",
    "    affine[0,0]/=4.0\n",
    "    affine[1,1]*=8.0\n",
    "    affine[2,2]/=4.0\n",
    "    print(affine)\n",
    "    new_image = nib.Nifti1Image(dummy.astype(np.uint8), affine=affine)\n",
    "    new_image.header[\"cal_min\"] = 0\n",
    "    new_image.header[\"cal_max\"] = 1\n",
    "    new_image.header[\"scl_slope\"] = 1/255.0\n",
    "    new_image.header[\"scl_inter\"] = 0\n",
    "    nib.save(new_image,workdir + \"../ref/new_ref_25mu.nii\")  # open in 3D Slicer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c04feaf4-6008-44aa-b873-330947ac7105",
   "metadata": {},
   "source": [
    "### 5. Align each gene by each RGB channel to BMCA-aligned backlit images"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84c471d-b963-4944-a92b-36cad0d4696d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 5. align each gene (move: sagittal stacks in 2) to the template backlit, save as niftis\n",
    "# output _2_STPT.nii.gz has repeating anterior and posterior slices\n",
    "\n",
    "def align_gene_2_bmcabl(raw_cmp,seg_cmp,mid):\n",
    "    print('------in 4 align_gene_2_bmcabl:')\n",
    "    #workdir_in = \"/disk/charissa/ISH_reg_pipeline/data/\"\n",
    "\n",
    "    jobids = []\n",
    "    \n",
    "    ref = workdir + \"../ref/new_ref_bl.nii\"\n",
    "    odir3 = workdir+mid+\"/out/01_backlit/3D/\" \n",
    "    POSTTRAFO = odir3+\"/trafo_01_\"\n",
    "    \n",
    "    for rcmp in raw_cmp:\n",
    "        if \"_gene\" in rcmp:\n",
    "            #SLURM_commands = []\n",
    "            for r in [\"img\",\"imgR\",\"imgG\",\"imgB\"]:\n",
    "                SLURM_commands = []\n",
    "                move = workdir+mid+\"/out/\"+rcmp+\"/3D/\"+r+\".nii.gz\"         # from sagittal stack (from 2)\n",
    "                OUT1 = workdir+mid+\"/out/\"+rcmp+\"/3D/\"+r+\"_2_STPT.nii.gz\"\n",
    "                logfolder = workdir+mid+\"/out/\"+rcmp+\"/log_3D_gene2bmcabl/\"\n",
    "                os.makedirs(logfolder,exist_ok=True)\n",
    "                cores = \"2\"\n",
    "                SLURM_commands += [\"MOVE=\"+move+\" REF=\"+ref+\" CORES=\"+cores+\" OUT1=\"+OUT1+\"  POSTTRAFO=\"+POSTTRAFO+\"  bash apply_avg3D_01.sh;\"]\n",
    "                #print('move: {}, ref: {}, OUT1: {}, POSTTRAFO: {}, logfolder: {}'.format(move, ref, OUT1, POSTTRAFO, logfolder))\n",
    "\n",
    "            jobid, success = slurm_tools.slurm_submit(SLURM_commands,\n",
    "                name = \"reg3D_rawgene2bmcabl\",\n",
    "                output = logfolder+'/log_01.out',\n",
    "                mem = '8GB',\n",
    "                cores = cores,\n",
    "                partition = \"ish-adult\", #\"bigmem\",\n",
    "                )      \n",
    "\n",
    "            print(\"job number: \"+jobid)\n",
    "\n",
    "            if not success:\n",
    "                    print(\"could not submit jobs\")\n",
    "                    print(format(jobid))\n",
    "                    print(format(jobids))\n",
    "                    slurm_tools.killall(jobids)\n",
    "            jobids.append(int(jobid))\n",
    "\n",
    "    slurm_tools.wait_for_jobs(jobids)\n",
    "    clear_output(wait=True)\n",
    "\n",
    "    for scmp in seg_cmp:\n",
    "        if \"_gene_seg\" in scmp:\n",
    "            #SLURM_commands = []\n",
    "            for s in [\"img\"]:\n",
    "                SLURM_commands = []\n",
    "                move = workdir+mid+\"/out_seg/\"+scmp+\"/3D/\"+s+\".nii.gz\"\n",
    "                OUT1 = workdir+mid+\"/out_seg/\"+scmp+\"/3D/\"+s+\"_2_STPT.nii.gz\"\n",
    "                logfolder = workdir+mid+\"/out_seg/\"+scmp+\"/log_3D_gene2bmcabl/\"\n",
    "                os.makedirs(logfolder,exist_ok=True)\n",
    "                cores = \"2\"\n",
    "                SLURM_commands += [\"MOVE=\"+move+\" REF=\"+ref+\" CORES=\"+cores+\" OUT1=\"+OUT1+\"  POSTTRAFO=\"+POSTTRAFO+\"  bash apply_avg3D_01.sh;\"]\n",
    "                #print('move: {}, ref: {}, OUT1: {}, POSTTRAFO: {}, logfolder: {}'.format(move, ref, OUT1, POSTTRAFO, logfolder))\n",
    "\n",
    "            jobid, success = slurm_tools.slurm_submit(SLURM_commands,\n",
    "                name = \"reg3D_seggene2bmcabl\",\n",
    "                output = logfolder+'/log_01.out',\n",
    "                mem = '8GB',\n",
    "                cores = cores,\n",
    "                partition = \"ish-adult\", #\"bigmem\",\n",
    "                )      \n",
    "\n",
    "            print(\"job number: \"+jobid)\n",
    "\n",
    "            if not success:\n",
    "                    print(\"could not submit jobs\")\n",
    "                    print(format(jobid))\n",
    "                    print(format(jobids))\n",
    "                    slurm_tools.killall(jobids)\n",
    "            jobids.append(int(jobid))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e9627e5-db22-4b2b-b7ea-8178e9cb4852",
   "metadata": {},
   "source": [
    "### 6. Save aligned backlit and gene niftis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4749655e-762b-420a-a272-745ac57bc012",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 6. make backlit and gene stacks individually as R/G/B and also the full RGB\n",
    "# aligned gene stacks are aligned to a lower res backlit template, because the original gene images are lower res\n",
    "# to be specific, they are 8-bit, and there are fewer z-slices than the template\n",
    "# so there is no reason to save them as higher bit\n",
    "\n",
    "def save_bl_gene_niis(raw_cmp,seg_cmp,mid):\n",
    "    print('------in 5 save_bl_gene_niis:')\n",
    "    ref = nib.load('/disk/charissa/ISH_reg_pipeline/ref/new_ref_bl.nii').get_fdata(dtype=np.float32)\n",
    "    z_mask = ref.max(axis=(0,2))  # (356,) consisting of 0. or 255.\n",
    "\n",
    "    ref2 = nib.load('/disk/charissa/ISH_reg_pipeline/ref/new_ref_25mu.nii').get_fdata(dtype=np.float32)\n",
    "    z_mask2 = ref2.max(axis=(0,2))  # (44,) consisting of 0. or 1.\n",
    "\n",
    "    all_out_3D = workdir+mid+\"/out/3D/low_res/\"\n",
    "    os.makedirs(all_out_3D,exist_ok=True)\n",
    "\n",
    "\n",
    "    for rcmp in raw_cmp:\n",
    "        #pass\n",
    "        if \"_backlit\" in rcmp or \"_gene\" in rcmp and \"seg\" not in rcmp:\n",
    "        #if \"_gene\" in cmp:    \n",
    "            ilist = [\"img\"] if \"_backlit\" in rcmp else  [\"imgR\",\"imgG\",\"imgB\"]\n",
    "\n",
    "            for il in ilist:\n",
    "            #if False:\n",
    "\n",
    "                OUT1 = workdir+mid+\"/out/\"+rcmp+\"/3D/\"+il+\"_2_STPT.nii.gz\"\n",
    "                #print('saving il: {} as OUT1: {}'.format(il, OUT1))\n",
    "                gimg_ = nib.load(OUT1)\n",
    "                gimg = gimg_.get_fdata(dtype=np.float32)  # (256, 356, 230)\n",
    "                #data = np.asanyarray(gimg.dataobj)\n",
    "                gimg[:,z_mask==0,:] = 0  # wherever the mask is black, in z, make it also black\n",
    "                if il == \"imgR\":\n",
    "                    shape = gimg.shape\n",
    "                    data8 = np.zeros(shape,dtype=np.dtype([('R', 'u1'), ('G', 'u1'), ('B', 'u1')]))  # Data-type with fields R,G,B, each being a uint8:\n",
    "                    #imgRGB = np.zeros([shape[0],shape[1]])\n",
    "                if il in [\"imgR\",\"imgG\",\"imgB\"]:\n",
    "                    data8[il.replace(\"img\",\"\")] = gimg*255\n",
    "\n",
    "\n",
    "                new_image = nib.Nifti1Image((gimg*255).astype(np.uint8), affine=gimg_.affine)\n",
    "                new_image.header[\"cal_min\"] = 0\n",
    "                new_image.header[\"cal_max\"] = 1\n",
    "                new_image.header[\"scl_slope\"] = 1/255.0\n",
    "                new_image.header[\"scl_inter\"] = 0\n",
    "                nib.save(new_image,OUT1.replace(\".nii.gz\",\"_8bit.nii.gz\")) \n",
    "                nib.save(new_image,all_out_3D+\"/\"+rcmp+\"_\"+il+\".nii.gz\") \n",
    "\n",
    "            if \"imgB\" in ilist:\n",
    "                new_image = nib.Nifti1Image(data8, affine=gimg_.affine)\n",
    "                #new_image.header[\"cal_min\"] = 0\n",
    "                #new_image.header[\"cal_max\"] = 1\n",
    "                #new_image.header[\"scl_slope\"] = 1/255.0\n",
    "                #new_image.header[\"scl_inter\"] = 0\n",
    "                nib.save(new_image,all_out_3D+\"/\"+rcmp+\"_RGB.nii.gz\") \n",
    "                print('saving RGB as '+all_out_3D+\"/\"+rcmp+\"_RGB.nii.gz\")\n",
    "\n",
    "    for scmp in seg_cmp:\n",
    "        if \"_gene_seg\" in scmp:\n",
    "        #if \"_gene\" in cmp:    \n",
    "            ilist = [\"img\"]\n",
    "\n",
    "            for il in ilist:\n",
    "            #if False:\n",
    "\n",
    "                OUT1 = workdir+mid+\"/out_seg/\"+scmp+\"/3D/\"+il+\"_2_STPT.nii.gz\"\n",
    "                #print('saving il: {} as OUT1: {}'.format(il, OUT1))\n",
    "                gimg_ = nib.load(OUT1)\n",
    "                gimg = gimg_.get_fdata(dtype=np.float32)  # (256, 356, 230)\n",
    "                #data = np.asanyarray(gimg.dataobj)\n",
    "                gimg[:,z_mask==0,:] = 0  # wherever the mask is black, in z, make it also black\n",
    "\n",
    "\n",
    "                new_image = nib.Nifti1Image((gimg*255).astype(np.uint8), affine=gimg_.affine)\n",
    "                #new_image.header[\"cal_min\"] = 0\n",
    "                #new_image.header[\"cal_max\"] = 1\n",
    "                #new_image.header[\"scl_slope\"] = 1/255.0\n",
    "                #new_image.header[\"scl_inter\"] = 0\n",
    "                nib.save(new_image,OUT1.replace(\".nii.gz\",\"_8bit.nii.gz\")) \n",
    "                nib.save(new_image,all_out_3D+\"/\"+scmp+\"_\"+il+\".nii.gz\") \n",
    "                print('saving seg as '+all_out_3D+\"/\"+scmp+\"_\"+il+\".nii.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61a9167-e51e-49d6-a09c-750a868835cd",
   "metadata": {},
   "source": [
    "# Calling all the functions and setting some global path variables #\n",
    "* where to find what\n",
    "* read in some basic info from meta (incorrect!!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06db9a04-cd08-4f3d-b034-2812c3ab2336",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = '/disk/charissa/shimogori_adult/'\n",
    "workdir = '/disk/charissa/ISH_reg_pipeline/data/'\n",
    "\n",
    "marm_ls = ['R08_0470']\n",
    "#['R04_0239','R08_0450','R08_0457','R08_0470','R08_0478','R08_0585','R08_0679','R08_0688','R08_0730']\n",
    "\n",
    "\n",
    "mu_bl = 3.12\n",
    "mu_bf = 28.72\n",
    "\n",
    "print(\"mu bl:\",mu_bl)\n",
    "print(\"mu bf:\",mu_bf)\n",
    "\n",
    "\n",
    "for d in range(len(marm_ls)):\n",
    "    mid = marm_ls[d]\n",
    "    print(mid)\n",
    "    \n",
    "    \n",
    "    jdata,success = json_read(db+mid+'/meta/01_backlit_000.json')\n",
    "    jdata,success = json_read(db+mid+'/meta/blockface_new/0108.json')\n",
    "    \n",
    "    raw_cmp = glob.glob(workdir+mid+\"/out/*\")\n",
    "    raw_cmp = [r.split(\"/\")[-1] for r in raw_cmp]\n",
    "    seg_cmp = glob.glob(workdir+mid+\"/out_seg/*\")\n",
    "    seg_cmp = [s.split(\"/\")[-1] for s in seg_cmp]\n",
    "    all_cmp = raw_cmp+seg_cmp    \n",
    "    \n",
    "\n",
    "    os.makedirs(workdir+mid+\"/bf/\",exist_ok=True)\n",
    "\n",
    "    sdir = db+mid+\"/img2d/\"\n",
    "\n",
    "    align_save_gene_rgbs(all_cmp, mid)  #1\n",
    "    make_sagittal_stacks(all_cmp,mid)    #2\n",
    "    align_bl_2_bmcabl(mid)               #3\n",
    "    align_gene_2_bmcabl(raw_cmp,seg_cmp,mid)    #5\n",
    "    save_bl_gene_niis(raw_cmp,seg_cmp,mid)      #6"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:jupyterhub] *",
   "language": "python",
   "name": "conda-env-jupyterhub-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
